File: 2022/week8/Programmers_92342.java
package week8;
import java.util.*;
public class Programmers_92342 {
    public static void main(String[] args) {
        new Solution().solution(5,new int[]{2,1,1,1,0,0,0,0,0,0,0});
//        new Solution().solution(1,new int[]{1,0,0,0,0,0,0,0,0,0,0});
//        new Solution().solution(9,new int[]{0,0,1,2,0,1,1,1,1,1,1});
//        new Solution().solution(10,new int[]{0,0,0,0,0,0,0,0,3,4,3});
    }
    static class Solution {
        int gap;
        int[] answer;
        List<Integer> nums;
        public int[] solution(int n, int[] info) {
            gap = 0;
            nums = new ArrayList<>();
            answer = new int[]{0,0,0,0,0,0,0,0,0,0,0};
            for(long i = 0 ; i < (long) Math.pow(10,11);i++){
                if(isValid(i,n)){
                    String strNum = Long.toString(i);
                    int size = strNum.length();
                    for(int j = 0 ; j < info.length-size; j++){
                        strNum="0"+strNum;
                    }
                    int apeach = 0;
                    int lion = 0;
                    for(int j = 0 ; j < info.length ;j++){
                        if(strNum.charAt(j)==0 && strNum.charAt(j)==info[j]) continue;
                        if(strNum.charAt(j)>info[j]){
                            lion+=(10-j);
                        }else apeach+=(10-j);
                    }
                    if(lion>apeach){
                        gap = Math.max(gap,lion-apeach);
                        for(int j = 0 ; j < info.length ;j++){
                            answer[j] = strNum.charAt(j);
                        }
                        System.out.println(Arrays.toString(answer));
                    }
                }
                i=i*10-1;
            }
            return gap == 0 ? new int[]{-1} : answer;
        }
        public boolean isValid(long num, int size){
            int sum = 0;
            while(num>0){
                sum+=(num%10);
                num/=10;
            }
            if(sum!=size) return false;
            return true;
        }
    }
}

File: 2022/week8/Programmers_92342.py
global gap_max

def solution(n, info):
    answer = [-1]
    bit = (1 << 12)-1
    gap_max = 0
    for i in range(bit,0):
        if is_valid(n,info,bit):
            return get_result(info, bit)
    return answer

def get_result(info, bit):
    arr = [0]*len(info)
    for i in range(len(info)):
        if bit & (1 << i) == (1 << i):
            arr[i] = info[i]+1


def is_valid(n, info, bit):
    lion = 0
    apeach = 0
    for i in range(len(info)):
        if bit & (1 << i) == (1 << i):
            n -= (info[i]+1)
            lion+=i
            if n < 0: return False
        elif bit & (1 << i):
            if info[i]!=0:
                apeach+=i

    gap_max = max([num,lion-apeach])
    return True


print(solution([2,1,1,1,0,0,0,0,0,0,0],[0,2,2,0,1,0,0,0,0,0,0]))